source/simulator/helper_functions.cc

```diff
@@ -909,8 +909,142 @@ namespace aspect
+
+            local_area_integrals=0;
+
+            for (unsigned int q=0; q<n_q_points_xy; ++q)
+              {
+                local_area_integrals+= fe_values_xy.JxW(q);
+              }
+            for (unsigned int c=0; c<parameters.n_compositional_fields; ++c)
+              {
+                fe_values_xy[introspection.extractors.compositional_fields[c]].get_function_values(vector,
+                    compositional_values_xy);
+                fe_values_yx[introspection.extractors.compositional_fields[c]].get_function_values(vector,
+                    compositional_values_yx);
+
+                local_compositional_average_integrals[c]=0;
+
+                for (unsigned int q=0; q<n_q_points_xy; ++q)
+                  {
+                    local_compositional_average_integrals[c]+= compositional_values_xy[q]*fe_values_xy.JxW(q);
+                  }
+                local_compositional_average_integrals[c] /= local_area_integrals;
+
+                min_composition=compositional_values_xy[0];
+                max_composition=compositional_values_xy[0];
+                for (unsigned int q=0; q<n_q_points_xy; ++q)
+                  {
+                    min_composition = std::min<double> (min_composition,
+                                                        compositional_values_xy[q]);
+                    max_composition = std::max<double> (max_composition,
+                                                        compositional_values_xy[q]);
+                  }
+                for (unsigned int q=0; q<n_q_points_yx; ++q)
+                  {
+                    min_composition = std::min<double> (min_composition,
+                                                        compositional_values_yx[q]);
+                    max_composition = std::max<double> (max_composition,
+                                                        compositional_values_yx[q]);
+                  }
+                // find the trouble cell
+                if (min_composition <C_min || max_composition >C_max)
+                  {
+                    //Define theta
+                    double theta_T=std::min<double>(1,abs((C_max-local_compositional_average_integrals[c])
+                                                          /(max_composition-local_compositional_average_integrals[c])));
+                    theta_T=std::min<double>(theta_T,abs((C_min-local_compositional_average_integrals[c])
+                                                         /(min_composition-local_compositional_average_integrals[c])));
+                    //Modify the numerical solution at freedoms
+                    std::vector<double> t_tmp(finite_element.base_element(introspection.base_elements.compositional_fields).dofs_per_cell);
+                    for (unsigned int j=0; j<finite_element.base_element(introspection.base_elements.compositional_fields).dofs_per_cell; ++j)
+                      {
+                        unsigned int support_point_index
```

Comment:
We usually declare everything 'const' if possible (to avoid uninteded changes later on), this would such a case.